[DropIn] Change a custom speed limit at runtime

kmadsen opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->

Issue in coordinators that observe customizable uibinders (e.g., `SpeedLimitCoordinator`). Essentially, the initial state is triggered before the `MapboxNavigationView.customize` function. To ensure this new state is applied on the `onStart` lifecycle, am adding an emit inside `override fun MapboxNavigation.flowViewBinders()`

So I was thinking, instead, lets make all uiBinders observable. This way it does not matter when you call MapboxNavitation.customize(uibinders). You will have a new view applied to the layout 🪂 

Minor issue when trying to build good transitions. For some reason transitions are not very good at restoring previous view. This issue is related https://github.com/mapbox/mapbox-navigation-android/issues/5516

### Screenshots or Gifs
<!-- Include media files to provide additional context. It's REALLY useful for UI related PRs (e.g. ![screenshot gif](link)) -->

https://user-images.githubusercontent.com/3021882/155818095-3d98e9e9-f934-431c-8430-3c161bbb3f73.mp4


<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
1. Adda a changelog entry under `Unreleased` tag or a `skip changelog` label if not applicable.
1. Update progress status on the project board.
1. Request a review from the team, if not a draft.
1. Add targeted milestone, when applicable.
1. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

tomaszrybakiewicz commented:
@kmadsen Looks like we forgot about this PR. @abhishek1508 can you review this?

abhishek1508 commented:
I looked at this PR but have some concerns. Waiting for @kmadsen to return and talk about it. Are you blocked by this PR in any way @tomaszrybakiewicz ?

tomaszrybakiewicz commented:
@abhishek1508 Got it. Thanks. No, I am not blocked. Just wanted to make sure we don't forget about this one.

kmadsen commented:
This is rebased and updated with `RoadNameLabelCoordinator` using the state flow. Also added an internal extension function that will help us make the coordinators consistent

``` kotlin
internal class SpeedLimitCoordinator {
--snip--
    override fun MapboxNavigation.flowViewBinders(): Flow<UIBinder> {
        return navigationViewContext.flowUiBinder { it.speedLimit }
    }
```

``` kotlin
internal class RoadNameLabelCoordinator {
--snip--
    override fun MapboxNavigation.flowViewBinders(): Flow<UIBinder> {
        return navigationViewContext.flowUiBinder { it.roadName }
    }
```

kmadsen commented:
I was playing around with transitions. The TransitionManager, or at least how we're using it, is not behaving exactly how I'd expect. So the fade or slide transitions do not work reliably. 

But I do consider this a good step forward, considering it gives us the ability to swap views at runtime.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewContext.kt:
```diff
@@ -26,3 +32,11 @@ internal class DropInNavigationViewContext(
+ */
+internal fun <T : UIBinder> DropInNavigationViewContext.flowUiBinder(
+    mapper: suspend (value: NavigationUIBinders) -> T
+): Flow<T> = this.uiBinders.map(mapper).distinctUntilChangedBy { it.javaClass }
```
I do this knowing that, new instances will not change the view. It needs to be a different class type.

I'm considering this better than, requiring all `UIBinder` implementations to override equals/hashcode.

abhishek1508 reviewed (approved):
Looks good 👍🏻
